File: truevfs-ext/truevfs-ext-logging/src/main/scala/net/java/truevfs/ext/logging/LogCloseable.scala
/*
 * Copyright (C) 2005-2015 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package net.java.truevfs.ext.logging

import java.io._
import net.java.truecommons.cio._

private trait LogCloseable extends Closeable with LogResource {

  log("Opened {}")

  abstract override def close(): Unit = {
    log("Closing {}")
    super.close
  }

  def log(message: String): Unit = {
    try {
      log(message, context.target())
    } catch {
      case ex: IOException =>
        logger trace ("Couldn't resolve resource target: ", ex)
    }
  }

  def context: IoSocket[_ <: Entry]
}

File: truevfs-ext/truevfs-ext-logging/src/main/java/net/java/truevfs/ext/logging/LogCloseable.java
package net.java.truevfs.ext.logging;

import net.java.truecommons.cio.Entry;
import net.java.truecommons.cio.IoSocket;

import java.io.Closeable;
import java.io.IOException;

interface LogCloseable extends Closeable, LogResource  {

    IoSocket<? extends Entry> context();

    default void opening() {
        log("Opening {}");
    }

    default void log(String message) {
        try {
            log(message, context().target());
        } catch (IOException e) {
            logger().trace("Couldn't resolve resource target: ", e);
        }
    }
}
